aws/client.go

```diff
@@ -0,0 +1,505 @@
+	}
+
+	keyPairName, err := c.EnsureKeyPair(ctx, params.KeyPairName)
+	if err != nil {
+		return out, fmt.Errorf("could not find a suitable key pair for a key: %w", err)
+	}
+
+	instanceType, err := c.EnsureInstanceType(ctx, params.InstanceType)
+	if err != nil {
+		return out, fmt.Errorf("could not find a suitable instance type: %w", err)
+	}
+
+	vpcID, err := c.EnsureSubnet(ctx, params.SubnetID)
+	if err != nil {
+		return out, fmt.Errorf("could not find a suitable subnet: %w", err)
+	}
+
+	securityGroupID, err := c.EnsureSecurityGroup(ctx, params.SecurityGroupName, vpcID)
+	if err != nil {
+		return out, fmt.Errorf("could not find a suitable security group: %w", err)
+	}
+
+	err = c.EnsureSecurityGroupIngressRules(ctx, securityGroupID)
+	if err != nil {
+		return out, fmt.Errorf("could not apply ingress security rules: %w", err)
+	}
+
+	params.UserData.CoreInstanceName = params.CoreInstanceName
+
+	userData, err := c.CreateUserdata(ctx, params.UserData)
+	if err != nil {
+		return out, fmt.Errorf("could not generate instance user data: %w", err)
+	}
+
+	runInstancesInput := &ec2.RunInstancesInput{
+		MaxCount:          aws.Int32(1),
+		MinCount:          aws.Int32(1),
+		ImageId:           aws.String(imageID),
+		InstanceType:      awstypes.InstanceType(instanceType),
+		UserData:          aws.String(userData),
+		SecurityGroupIds:  []string{securityGroupID},
+		SubnetId:          aws.String(params.SubnetID),
+		KeyName:           aws.String(keyPairName),
+		TagSpecifications: c.getTags(awstypes.ResourceTypeInstance),
+	}
+
+	instances, err := c.client.RunInstances(ctx, runInstancesInput)
+	if err != nil {
+		return out, err
+	}
+
+	instance := instances.Instances[0]
+	out.EC2InstanceID = *instance.InstanceId
+	out.EC2InstanceType = string(instance.InstanceType)
+	out.AMIID = *instance.ImageId
+	out.Hostname = *instance.PublicDnsName
+	out.PrivateIPv4 = *instance.PrivateIpAddress
+
+	// await for the instance to reach available status, cannot be associated with an
+	// IPv4 address if the instance is not ready.
+	if params.PublicIPAddress != nil {
+		err = retry.Do(ctx, instanceUpCheckMaxDuration(), func(ctx context.Context) error {
+			state, err := c.InstanceState(ctx, *instance.InstanceId)
+			if err != nil {
+				return retry.RetryableError(err)
+			}
+
+			if state != string(awstypes.InstanceStateNameRunning) {
+				return retry.RetryableError(fmt.Errorf("instance not in running state"))
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			return out, err
+		}
+
+		publicIPv4Address, err := c.EnsureAndAssociateElasticIPv4Address(ctx, out.EC2InstanceID,
+			params.PublicIPAddress.Pool, params.PublicIPAddress.Address)
+		if err != nil {
+			return out, fmt.Errorf("could not associate public ipv4 address: %w", err)
+		}
+		out.PublicIPv4 = publicIPv4Address
+	}
+
+	out.CoreInstanceName = params.CoreInstanceName
+	return out, err
```

Comment:
```suggestion
	if params.PublicIPAddress == nil {
		out.CoreInstanceName = params.CoreInstanceName
	        return out, err
	}
err = retry.Do(ctx, instanceUpCheckMaxDuration(), func(ctx context.Context) error {
	state, err := c.InstanceState(ctx, *instance.InstanceId)
	if err != nil {
		return retry.RetryableError(err)
	}

	if state != string(awstypes.InstanceStateNameRunning) {
		return retry.RetryableError(fmt.Errorf("instance not in running state"))
	}

	return nil
})

if err != nil {
	return out, err
}

publicIPv4Address, err := c.EnsureAndAssociateElasticIPv4Address(ctx, out.EC2InstanceID,
	params.PublicIPAddress.Pool, params.PublicIPAddress.Address)
if err != nil {
	return out, fmt.Errorf("could not associate public ipv4 address: %w", err)
}
out.PublicIPv4 = publicIPv4Address
```

maybe we can invert this nil verify and keep the nest short